<a href="https://colab.research.google.com/github/felipesayegg/Previs-o-de-Pre-os-de-Ve-culos/blob/main/RandonFlorestAuto.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Estrutura do Documento

1. Introdução

Nesta sessão, introduzimos o propósito do projeto: Previsão de preços de veículos utilizando técnicas de aprendizado de máquina, com ênfase no algoritmo Random Forest.

 2. Coleta de Dados

Descrição do método de coleta de dados: Fonte dos dados, formato e possíveis desafios na obtenção e preparação inicial.

 3. Análise de Dados e Tratamento

- **3.1 Análise de Variáveis:**
  - Avaliação inicial para entender os dados disponíveis, incluindo tipos de variáveis e possíveis relações.

- **3.2 Análise Descritiva:**
  - Estatísticas básicas como média, mediana, desvio padrão, etc., para obter insights iniciais sobre a dispersão e centralização dos dados.

- **3.3 Tratando Outliers:**
  - Identificação e ajustes em valores extremos para evitar distorções nas análises subsequentes.

- **3.4 Tratamento de Valores Faltantes:**
  - Estratégias para lidar com dados ausentes, seja removendo ou imputando esses valores.

4. Pré-Processamento, Modelagem e Previsão

- **4.1 Pré-Processamento  (OneHot Encoding, Escalonamento e Variáveis Numéricas):**
  - **OneHot Encoding:** Transformação de variáveis categóricas em formato binário.
  - **Escalonamento:** Normalização/padronização das variáveis numéricas para consistência nos dados.
  - **Pré-Processamento de Variáveis Numéricas:**
    - Aplicação de normalização, padronização, tratamento de outliers e transformações numéricas como log e root, além da imputação de valores faltantes.

- **4.2 Divisão de Dados (Treinamento/Teste):**
  - Separação dos dados em conjuntos de treinamento e teste para validação do modelo.

- **4.3 Treinamento do Modelo Random Forest:**
  - Implementação e ajuste do modelo Random Forest usando o conjunto de dados de treinamento.

- **4.4 Avaliação de Desempenho do Modelo:**
  - Avaliação da eficácia do modelo utilizando métricas de desempenho apropriadas.

 5. Conclusão

Revisão dos resultados obtidos, discussões sobre as implicações dos achados e sugestões para trabalhos futuros.

# 1  Introdução - Previsão de Preços de Veículos com Random Forest  

## 📌 Desvendando o Mercado Automotivo  

Prever com precisão o preço de veículos é um desafio essencial no mercado automotivo, onde diversos fatores como quilometragem, potência do motor e tipo de combustível influenciam diretamente o valor final de um carro. Com um conjunto de dados extenso e diversificado, a tarefa exige uma abordagem robusta para lidar com a complexidade das variáveis e fornecer estimativas precisas.  

## 🎯 Problema de Regressão  

Este projeto tem como foco a **previsão dos preços de veículos** utilizando variáveis numéricas e categóricas que capturam diferentes aspectos dos automóveis. Como os preços são valores contínuos e influenciados por múltiplos fatores interdependentes, este problema se enquadra em uma **tarefa de regressão** no campo do aprendizado de máquina.  

## 💡 Solução Proposta  

Para lidar com esse desafio, utilizaremos o **Random Forest Regressor**, um modelo de aprendizado de máquina baseado em múltiplas árvores de decisão. Essa abordagem é amplamente reconhecida por sua capacidade de capturar padrões não lineares, reduzir o risco de overfitting e gerar previsões mais estáveis e confiáveis, especialmente em grandes volumes de dados.  

## 🎯 Objetivos do Projeto  

✅ Desenvolver um modelo preditivo capaz de estimar com precisão os preços dos veículos com base em suas características.  
✅ Compreender as interações entre as variáveis mais relevantes para identificar os fatores que mais impactam o valor de um carro.  
✅ Fornecer insights valiosos para consumidores, vendedores e analistas do setor automotivo.  

🚀 **Vamos construir um modelo eficiente e entender melhor o mercado de veículos!**  




# 2 - Coleta de dados

In [330]:

import pandas as pd                    # Importando a biblioteca Pandas, que é como uma planilha turbinada para analisar dados
import numpy as np                     # Importando a biblioteca NumPy, que ajuda a fazer contas e operações com números de forma eficiente
import matplotlib.pyplot as plt        # Importando a biblioteca Matplotlib, que serve para criar gráficos e visualizações
import seaborn as sns                  # Importando a biblioteca Seaborn, que deixa os gráficos mais bonitos e informativos
from sklearn.model_selection import train_test_split, GridSearchCV  # Importando ferramentas para dividir os dados e encontrar as melhores configurações do modelo
from sklearn.preprocessing import StandardScaler, LabelEncoder, OneHotEncoder  # Importando ferramentas para preparar os dados para o modelo
from sklearn.ensemble import RandomForestRegressor  # Importando o modelo de Random Forest, que vamos usar para prever os preços
from sklearn.metrics import mean_absolute_error     # Importando a ferramenta para medir o quão bom o modelo está
import missingno as msno               # Importando uma biblioteca para visualizar dados faltantes (opcional)
import datetime                        # Importando a biblioteca para trabalhar com datas e horários
from datetime import datetime
# Importando a biblioteca para trabalhar com datas e horários.

In [331]:
dados = pd.read_csv('/content/autos.csv', encoding='ISO-8859-1', sep=',')
# Carregando o arquivo CSV 'autos.csv' que contém os dados dos carros.
# O encoding='ISO-8859-1' é para garantir que os caracteres especiais sejam lidos corretamente.
# O sep=',' indica que os valores no arquivo são separados por vírgulas.

In [332]:
dados.head() # Mostrar as 5 primeiras linhas do meu dataset

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

# 3 - Analise de dados e Tratamento

### 3.1 - Analise de variaveis

In [333]:
colunas_para_remover = [
    'dateCrawled', 'name', 'seller', 'offerType', 'dateCreated',
    'nrOfPictures', 'postalCode', 'lastSeen', 'monthOfRegistration'
]
# Criando uma lista com os nomes das colunas que não vamos usar no modelo.
# Essas colunas podem ter informações que não ajudam a prever o preço do carro.


In [334]:
dados = dados.drop(columns=colunas_para_remover, errors='ignore')
# Removendo as colunas da lista do DataFrame.
# O errors='ignore' faz com que o código ignore se alguma coluna não existir.

In [335]:
dados.head() # Mostrando as primeiras linhas do arquivo para ver como ficou depois de remover as colunas

price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer fuelType       brand notRepairedDamage  
0     150000   benzin  volkswagen               NaN  
1     125000   diesel        audi                ja  
2     125000   diesel        jeep               NaN  
3     150000   benzin  volkswagen              nein  
4      90000   diesel       skoda              nein

Neste modelo de previsão de preço de veículos, mantivemos as seguintes variáveis, cada uma selecionada por sua contribuição significativa para prever o preço final do veículo:

- **price**: Variável alvo. Preço que estamos tentando prever.
- **abtest**: Indica a participação em testes A/B, podendo evidenciar variações na eficiência do marketing.
- **vehicleType**: O tipo do veículo (SUV, sedan, etc.) afeta diretamente o valor de mercado.
- **yearOfRegistration**: Ano de fabricação do veículo; veículos mais novos geralmente têm preços mais altos.
- **gearbox**: Tipo de câmbio (manual ou automático), que influenciam o preço devido a preferências regionais/culturais.
- **powerPS**: Potência do motor; motores mais potentes costumam ter preços mais elevados.
- **model**: O modelo pode variar bastante em preço dentro da mesma marca devido a características específicas.
- **kilometer**: Quilometragem rodada; menor quilometragem usualmente implica em maior valor.
- **fuelType**: Tipo de combustível (gasolina, diesel, etc.), que pode afetar o preço devido a custos operacionais.
- **brand**: Marca do veículo; marcas com maior prestígio frequentemente têm valores mais elevados.
- **notRepairedDamage**: Indica danos não reparados, podendo diminuir significativamente o valor de um veículo.

Essas variáveis foram escolhidas por sua relevância demonstrada na literatura e prática de modelagem preditiva de preços, capturando fatores críticos que influenciam a avaliação dos preços de veículos.

### 3.2 - Analise Descritiva

In [336]:
dados.describe() # Calculando estatísticas básicas sobre os dados, como média, mediana, desvio padrão, etc.
# Isso ajuda a entender como os dados estão distribuídos e a identificar possíveis problemas.

price  yearOfRegistration        powerPS      kilometer
count  3.715280e+05       371528.000000  371528.000000  371528.000000
mean   1.729514e+04         2004.577997     115.549477  125618.688228
std    3.587954e+06           92.866598     192.139578   40112.337051
min    0.000000e+00         1000.000000       0.000000    5000.000000
25%    1.150000e+03         1999.000000      70.000000  125000.000000
50%    2.950000e+03         2003.000000     105.000000  150000.000000
75%    7.200000e+03         2008.000000     150.000000  150000.000000
max    2.147484e+09         9999.000000   20000.000000  150000.000000

Estatísticas Descritivas das Variáveis Numéricas

Abaixo estão as estatísticas descritivas principais sobre as variáveis numéricas do conjunto de dados, obtidas através do método `describe()`:

Explicação dos Termos:
- **count**: Quantidade total de valores não nulos na variável.
- **mean**: Média aritmética dos valores.
- **std**: Desvio padrão, que mede a dispersão dos dados ao redor da média.
- **min**: Valor mínimo encontrado.
- **25%**: Percentil 25, indicando o valor abaixo do qual estão 25% dos dados.
- **50%** (mediana): Valor no meio do conjunto de dados (50% acima e 50% abaixo).
- **75%**: Percentil 75, indicando o valor abaixo do qual estão 75% dos dados.
- **max**: Valor máximo encontrado.

Resumo por Variável

1. **price** (Preço do veículo - variável alvo)
- **count**: 371,528 registros.
- **mean**: 17,291.4 unidades monetárias (valor médio).
- **std**: 3,587,954.6, indicando alta dispersão, possivelmente devido a valores outliers (veículos com preços extremamente altos).
- **min**: 0, indicando que alguns preços podem estar incorretos (exemplo de dados inválidos).
- **25%**: 1,150.
- **50% (mediana)**: 2,950.
- **75%**: 7,200.
- **max**: 2,147,483,647, um valor extremamente alto que pode ser um erro.

---

 2. **yearOfRegistration** (Ano de Registro do veículo)
- **count**: 371,528 registros.
- **mean**: 2004.58, sugerindo que a maioria dos veículos é relativamente recente (pós-2000).
- **std**: 92.87, indicando uma dispersão incomum, provavelmente devido a alguns valores de ano totalmente inválidos.
- **min**: 1000, claramente um valor errado.
- **25%**: 1999.
- **50% (mediana)**: 2003.
- **75%**: 2008.
- **max**: 9999, valor inválido.

---
3. **powerPS** (Potência do veículo, em cavalos)
- **count**: 371,528 registros.
- **mean**: 115.55 cv (potência média).
- **std**: 192.14, indicando dispersão significativa, possivelmente devido a valores anômalos (muito altos ou muito baixos).
- **min**: 0 cv, o que é inválido para veículos funcionais.
- **25%**: 70 cv.
- **50% (mediana)**: 105 cv.
- **75%**: 150 cv.
- **max**: 20,000 cv, um valor claramente irreal.

---

4. **kilometer** (Quilometragem percorrida)
- **count**: 371,528 registros.
- **mean**: 125,618.7 km (média de distância percorrida).
- **std**: 40,112.3, indicando dispersão razoável para quilometragem.
- **min**: 5,000 km, representando veículos praticamente novos.
- **25%**: 125,000 km.
- **50% (mediana)**: 150,000 km.
- **75%**: 150,000 km.
- **max**: 150,000 km, sugerindo que os valores podem ser truncados, com muitas observações concentradas na mesma faixa.

---

Conclusões
- Algumas variáveis apresentam **valores inválidos ou discrepantes**, como `yearOfRegistration` (anos fora do intervalo esperado), `powerPS` (potências irrealisticamente altas) e `price` (valores extremamente altos ou zero).
- Esses **outliers e valores inválidos podem ser tratados** antes do treinamento do modelo para evitar impacto negativo nos resultados.

### 3.3 - Tratando outliers

In [337]:
dados.loc[dados['price'] <= 100] # Verificando quantos dados tem com o valor a baixo de 100

price   abtest vehicleType  yearOfRegistration  gearbox  powerPS  \
7           0     test   limousine                1980  manuell       50   
40          0     test         NaN                1990      NaN        0   
60          1  control         suv                1994  manuell      286   
91          1  control   limousine                1995  manuell      113   
115         0     test         NaN                2017  manuell        0   
...       ...      ...         ...                 ...      ...      ...   
371392      0     test  kleinwagen                2002  manuell       60   
371402      0  control  kleinwagen                1999  manuell       53   
371431      0  control  kleinwagen                1999  manuell       37   
371464     99     test  kleinwagen                1999  manuell       60   
371522      0  control         NaN                2005  manuell        0   

           model  kilometer fuelType           brand notRepairedDamage  
7         andere      40000   benzin      volkswagen              nein  
40         corsa     150000   benzin            opel               NaN  
60           NaN     150000      NaN  sonstige_autos               NaN  
91      e_klasse     150000   diesel   mercedes_benz              nein  
115         golf       5000   benzin      volkswagen               NaN  
...          ...        ...      ...             ...               ...  
371392    fiesta     150000   benzin            ford               NaN  
371402     swift     150000   benzin          suzuki               NaN  
371431     arosa     150000   benzin            seat                ja  
371464     arosa     150000   benzin            seat                ja  
371522      colt     150000   benzin      mitsubishi                ja  

[14352 rows x 11 columns]

In [338]:
dados.loc[dados['price'] >= 350000] # verificando quantos dados tem com valor acima de 350000

price   abtest vehicleType  yearOfRegistration  gearbox  powerPS  \
1846      579000  control       coupe                1980  manuell      277   
10649     420000  control       coupe                2004  manuell      483   
14663   11111111  control       coupe                2003  manuell       64   
16889    1000000  control       kombi                1998      NaN        0   
20143    1250000     test       coupe                2016  manuell      500   
...          ...      ...         ...                 ...      ...      ...   
364171   3890000     test       coupe                2006      NaN      799   
365461    599000  control       coupe                1980  manuell      377   
366653  99999999  control      cabrio                1996  manuell      192   
366861   3895000     test       coupe                2006      NaN      799   
367041   1600000     test       coupe                2005      NaN      625   

         model  kilometer fuelType           brand notRepairedDamage  
1846    andere      20000   benzin             bmw              nein  
10649      911      50000   benzin         porsche              nein  
14663     polo     150000   benzin      volkswagen               NaN  
16889   mondeo     150000   benzin            ford                ja  
20143      911       5000   benzin         porsche              nein  
...        ...        ...      ...             ...               ...  
364171     NaN       5000      NaN  sonstige_autos              nein  
365461  andere       5000   benzin             bmw              nein  
366653     3er     150000      NaN             bmw               NaN  
366861     NaN       5000   benzin  sonstige_autos              nein  
367041     NaN      10000   benzin  sonstige_autos              nein  

[119 rows x 11 columns]

In [339]:
mediana_preco = dados["price"].median()
mediana_preco

2950.0

In [340]:
# Substituindo valores de preço considerados muito baixos (abaixo de 100) pela mediana
dados.loc[dados['price'] < 100, 'price'] = mediana_preco

In [341]:
# Substituindo valores de preço considerados muito altos (acima de 350000) pela mediana
dados.loc[dados['price'] > 350000, 'price'] = mediana_preco

In [342]:
dados["price"].mean() # Calculando a média dos preços após tratamento

np.float64(5832.207995090545)

In [343]:
dados.describe() # Analise descritiva dos dados agora com price tratado tirando outlier

price  yearOfRegistration        powerPS      kilometer
count  371528.000000       371528.000000  371528.000000  371528.000000
mean     5832.207995         2004.577997     115.549477  125618.688228
std      8739.660684           92.866598     192.139578   40112.337051
min       100.000000         1000.000000       0.000000    5000.000000
25%      1300.000000         1999.000000      70.000000  125000.000000
50%      2950.000000         2003.000000     105.000000  150000.000000
75%      7200.000000         2008.000000     150.000000  150000.000000
max    350000.000000         9999.000000   20000.000000  150000.000000

### 3.4 - Tratamento Valores faltantes

In [344]:
dados.info()
# Verificando informações sobre os dados, como o tipo de cada coluna e quantos valores não nulos existem.


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 11 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   price               371528 non-null  int64 
 1   abtest              371528 non-null  object
 2   vehicleType         333659 non-null  object
 3   yearOfRegistration  371528 non-null  int64 
 4   gearbox             351319 non-null  object
 5   powerPS             371528 non-null  int64 
 6   model               351044 non-null  object
 7   kilometer           371528 non-null  int64 
 8   fuelType            338142 non-null  object
 9   brand               371528 non-null  object
 10  notRepairedDamage   299468 non-null  object
dtypes: int64(4), object(7)
memory usage: 31.2+ MB


In [345]:
print(dados.isnull().sum())
# Contando quantos valores nulos (faltantes) existem em cada coluna.

price                     0
abtest                    0
vehicleType           37869
yearOfRegistration        0
gearbox               20209
powerPS                   0
model                 20484
kilometer                 0
fuelType              33386
brand                     0
notRepairedDamage     72060
dtype: int64


In [346]:
moda_vehicle_type = dados['vehicleType'].mode()[0]
# Calculando a moda (valor mais frequente) da coluna 'vehicleType'.

In [347]:
# Exibe a moda calculada para verificar o resultado
print(f"A moda da coluna 'vehicleType' é: {moda_vehicle_type}")

A moda da coluna 'vehicleType' é: limousine


In [348]:
dados['vehicleType'] = dados['vehicleType'].fillna(moda_vehicle_type)
# Preenchendo os valores nulos da coluna 'vehicleType' com a moda.

In [349]:
# Verifica se ainda há valores nulos após a substituição
print(f"Valores nulos restantes na coluna 'vehicleType': {dados['vehicleType'].isnull().sum()}")

Valores nulos restantes na coluna 'vehicleType': 0


In [350]:
dados['gearbox'].value_counts()
# Contando quantos valores diferentes existem na coluna 'gearbox'.

gearbox
manuell      274214
automatik     77105
Name: count, dtype: int64

In [351]:
moda_gearbox = dados['gearbox'].mode()[0]
# Calculando a moda da coluna 'gearbox'.

# Exibe a moda calculada para verificar qual é o valor mais comum
print(f"A moda da coluna 'gearbox' é: {moda_gearbox}")

A moda da coluna 'gearbox' é: manuell


In [352]:
# Substitui valores nulos em 'gearbox' pela moda calculada
dados['gearbox'] = dados['gearbox'].fillna(moda_gearbox)

# Verifica se ainda existem valores nulos após a substituição
print(f"Valores nulos restantes na coluna 'gearbox': {dados['gearbox'].isnull().sum()}")

Valores nulos restantes na coluna 'gearbox': 0


In [353]:
dados['model'] = dados['model'].fillna('Desconhecido')
# Preenchendo os valores nulos da coluna 'model' com a string 'Desconhecido'

In [354]:
dados['model'].value_counts()
# Contando quantos valores diferentes existem na coluna 'model'.

model
golf               30070
andere             26400
3er                20567
Desconhecido       20484
polo               13092
                   ...  
serie_2                8
rangerover             6
serie_3                4
serie_1                2
discovery_sport        1
Name: count, Length: 252, dtype: int64

In [355]:
moda_fueltype = dados['fuelType'].mode()[0]
# Calculando a moda da coluna 'fuelType'.

dados['fuelType'] = dados['fuelType'].fillna(moda_fueltype)
# Preenchendo os valores nulos da coluna 'fuelType' com a moda.
# Verifica se restaram nulos
print(f"Valores nulos em 'fuelType': {dados['fuelType'].isnull().sum()}")

Valores nulos em 'fuelType': 0


In [356]:
# Contagem de valores únicos na coluna "notRepairedDamage"
dados['notRepairedDamage'].value_counts()
# Isso exibe quantos carros têm 'nein', 'ja' ou estão sem informação (NaN).

notRepairedDamage
nein    263182
ja       36286
Name: count, dtype: int64

In [357]:
# Definindo os valores padrão para preencher os valores ausentes de cada coluna
valores = {
    'model': 'golf',             # Preencher valores ausentes com 'golf' (modelo mais frequente ou outro pré-definido)
    'fuelType': 'benzin',        # Preencher valores ausentes com 'benzin' (gasolina)
    'notRepairedDamage': 'nein'  # Preencher valores ausentes com 'nein' (não tem danos reparados)
}

In [358]:
# Preenchendo valores ausentes nas colunas especificadas
dados.fillna(value=valores, inplace=True)
# O método fillna preenche os valores ausentes (NaN) com os dados definidos no dicionário 'valores'.

In [359]:
print("Valores nulos por coluna (antes do One-Hot Encoding):")
print(dados.isnull().sum())
# Verificando se ainda existem valores nulos no DataFrame.

Valores nulos por coluna (antes do One-Hot Encoding):
price                 0
abtest                0
vehicleType           0
yearOfRegistration    0
gearbox               0
powerPS               0
model                 0
kilometer             0
fuelType              0
brand                 0
notRepairedDamage     0
dtype: int64


# 4 - Pré-Processamento, Modelagem e Previsão

###   4.1 - Pré-Processamento (OneHot Encoding e Escalonamento)

Pré-Processamento Categorico

In [360]:
categorical_columns = ['abtest', 'vehicleType', 'gearbox', 'model', 'fuelType', 'brand', 'notRepairedDamage']
# Criando uma lista com os nomes das colunas que são categorias (textos).

In [361]:
encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
# Criando um objeto OneHotEncoder para transformar as colunas de texto em números.
# O sparse_output=False faz com que o resultado seja uma matriz densa (mais fácil de usar).
# O handle_unknown='ignore' faz com que o código ignore categorias desconhecidas.

In [362]:

encoded_data = encoder.fit_transform(dados[categorical_columns])
# Transformando as colunas de texto em números usando o OneHotEncoder.

In [363]:
print(f"Shape dos dados codificados: {encoded_data.shape}")
# Imprimindo o formato dos dados transformados.

Shape dos dados codificados: (371528, 313)


In [364]:
colunas_codificadas = encoder.get_feature_names_out(categorical_columns)
# Obtendo os nomes das novas colunas criadas pelo OneHotEncoder.

In [365]:
print(f"Número de colunas geradas de codificação: {len(colunas_codificadas)}")
# Imprimindo o número de colunas criadas.

Número de colunas geradas de codificação: 313


In [366]:
encoded_df = pd.DataFrame(encoded_data, columns=colunas_codificadas)
# Criando um DataFrame com os dados transformados.

In [367]:
encoded_df.index = dados.index
# Definindo o índice do novo DataFrame para que seja o mesmo do DataFrame original.

In [368]:
df_final = pd.concat([dados, encoded_df], axis=1)
# Juntando o DataFrame original com o DataFrame transformado.

In [369]:
df_final.drop(columns=categorical_columns, inplace=True)
# Removendo as colunas originais de texto.

In [370]:

print("Status de valores nulos após a concatenação final:")
print(df_final.isnull().sum().sum())
# Verificando se ainda existem valores nulos no DataFrame.

Status de valores nulos após a concatenação final:
0


In [371]:
nulos = df_final[df_final.isnull().any(axis=1)]
# Criando um DataFrame com as linhas que contêm valores nulos.

In [372]:
nulos.head(30)
# Mostrando as primeiras linhas do DataFrame com valores nulos.

Empty DataFrame
Columns: [price, yearOfRegistration, powerPS, kilometer, abtest_control, abtest_test, vehicleType_andere, vehicleType_bus, vehicleType_cabrio, vehicleType_coupe, vehicleType_kleinwagen, vehicleType_kombi, vehicleType_limousine, vehicleType_suv, gearbox_automatik, gearbox_manuell, model_100, model_145, model_147, model_156, model_159, model_1_reihe, model_1er, model_200, model_2_reihe, model_300c, model_3_reihe, model_3er, model_4_reihe, model_500, model_5_reihe, model_5er, model_601, model_6_reihe, model_6er, model_7er, model_80, model_850, model_90, model_900, model_9000, model_911, model_Desconhecido, model_a1, model_a2, model_a3, model_a4, model_a5, model_a6, model_a8, model_a_klasse, model_accord, model_agila, model_alhambra, model_almera, model_altea, model_amarok, model_andere, model_antara, model_arosa, model_astra, model_auris, model_avensis, model_aveo, model_aygo, model_b_klasse, model_b_max, model_beetle, model_berlingo, model_bora, model_boxster, model_bravo, model_c1, model_c2, model_c3, model_c4, model_c5, model_c_klasse, model_c_max, model_c_reihe, model_caddy, model_calibra, model_captiva, model_carisma, model_carnival, model_cayenne, model_cc, model_ceed, model_charade, model_cherokee, model_citigo, model_civic, model_cl, model_clio, model_clk, model_clubman, model_colt, model_combo, model_cooper, model_cordoba, ...]
Index: []

[0 rows x 317 columns]

Pré-Processamento Numerico

In [373]:
ano = datetime.now().year
# Obtendo o ano atual.

In [374]:

df_final["idade_veiculo"] = ano - df_final["yearOfRegistration"]
# Criando uma coluna com a idade do veículo.

In [375]:

df_final.drop("yearOfRegistration", axis=1, inplace=True)
# Removendo a coluna original com o ano de registro.

In [376]:
columns_to_scale = ['powerPS', 'kilometer', 'idade_veiculo']
# Criando uma lista com os nomes das colunas que serão padronizadas.


In [377]:
padronizacao = StandardScaler()
# Criando um objeto StandardScaler para padronizar os dados.

In [378]:
scaled_values = padronizacao.fit_transform(df_final[columns_to_scale])
# Padronizando as colunas selecionadas.

In [379]:
df_final[columns_to_scale] = scaled_values
# Substituindo as colunas originais pelas colunas padronizadas.

In [380]:
print(df_final.head())
# Mostrando as primeiras linhas do DataFrame para ver como ficou depois de padronizar os dados.

   price   powerPS  kilometer  abtest_control  abtest_test  \
0    480 -0.601384   0.607827             0.0          1.0   
1  18300  0.387482  -0.015424             0.0          1.0   
2   9800  0.246959  -0.015424             0.0          1.0   
3   1500 -0.211042   0.607827             0.0          1.0   
4   3600 -0.242269  -0.887975             0.0          1.0   

   vehicleType_andere  vehicleType_bus  vehicleType_cabrio  vehicleType_coupe  \
0                 0.0              0.0                 0.0                0.0   
1                 0.0              0.0                 0.0                1.0   
2                 0.0              0.0                 0.0                0.0   
3                 0.0              0.0                 0.0                0.0   
4                 0.0              0.0                 0.0                0.0   

   vehicleType_kleinwagen  ...  brand_sonstige_autos  brand_subaru  \
0                     0.0  ...                   0.0           0.0   


### 4.2 - Divisão de Dados (Treinamento/Teste)

In [381]:
X = df_final.drop(columns=['price'])
# Criando uma matriz com as colunas que serão usadas para prever o preço.


In [382]:
y = df_final['price']
# Criando um vetor com os preços dos carros.

In [383]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# Dividindo os dados em dois conjuntos: um para treinar o modelo e outro para testar o modelo.
# O test_size=0.2 indica que 20% dos dados serão usados para teste.
# O random_state=42 garante que a divisão seja sempre a mesma.

In [384]:
print("Tamanho do conjunto de treinamento:", X_train.shape)
# Imprimindo o formato do conjunto de treinamento.

Tamanho do conjunto de treinamento: (297222, 316)


In [385]:
print("Tamanho do conjunto de teste:", X_test.shape)
# Imprimindo o formato do conjunto de teste.

Tamanho do conjunto de teste: (74306, 316)


### 4.3 Treinamento do Modelo Random Forest:

In [400]:
# Definindo um conjunto de parâmetros simples para acelerar o treinamento
param_grid = {
    'n_estimators': [40, 60],  # Número de árvores na floresta. Aqui estamos testando apenas 2 valores
    'max_depth': [5, 10],       # Profundidade máxima de cada árvore. Testaremos árvores mais rasas para reduzir custos computacionais
    'min_samples_split': [2, 5] # Número mínimo de amostras necessárias para dividir um nó
}


In [401]:
grid_search = GridSearchCV(
    estimator=RandomForestRegressor(random_state=42),  # Modelo Random Forest com uma semente fixa para reprodutibilidade
    param_grid=param_grid,                             # Passando o grid de parâmetros configurado acima
    cv=3,                                              # Validação cruzada com 3 folds (divisões dos dados para validação)
    scoring='neg_mean_absolute_error',                 # Métrica para medir o desempenho (neste caso, MAE - erro absoluto)
    verbose=2,                                         # Mostra mensagens detalhadas no console durante a execução
    n_jobs=-1                                          # Utiliza todos os núcleos disponíveis no hardware para acelerar o cálculo
)


In [402]:
# Ajustando o modelo aos dados de treinamento
grid_search.fit(X_train, y_train)
# O GridSearchCV irá testar todas as combinações de parâmetros acima, calculando os resultados para cada combinação usando validação cruzada

Fitting 3 folds for each of 8 candidates, totalling 24 fits


GridSearchCV(cv=3, estimator=RandomForestRegressor(random_state=42), n_jobs=-1,
             param_grid={'max_depth': [5, 10], 'min_samples_split': [2, 5],
                         'n_estimators': [40, 60]},
             scoring='neg_mean_absolute_error', verbose=2)

In [403]:
# Exibindo os melhores hiperparâmetros encontrados
print("Melhores hiperparâmetros:", grid_search.best_params_)


Melhores hiperparâmetros: {'max_depth': 10, 'min_samples_split': 2, 'n_estimators': 60}


In [404]:
# Selecionando o melhor modelo encontrado pelo GridSearchCV
modelo_rf = grid_search.best_estimator_
# Aqui, grid_search.best_estimator_ pega o modelo com a melhor combinação de parâmetros para futuros testes

4.4 Avaliação de Desempenho do Modelo:

In [405]:
# Realiza as previsões com o modelo treinado utilizando os dados de teste
y_pred = modelo_rf.predict(X_test)


In [406]:
# Calcula o Mean Absolute Error (MAE) para medir o erro médio absoluto
mae = mean_absolute_error(y_test, y_pred)

In [407]:
# Exibe os resultados das métricas
print("Mean Absolute Error (MAE):", mae)

Mean Absolute Error (MAE): 1809.8759114730908
